### PR TITLE
feat(protocol): Add airdrop contracts

### PR DESCRIPTION
**Pull request by adaki2004**
Same as here: https://github.com/taikoxyz/taiko-mono/pull/15122 but that one had lots of conflicts so rebased to the new base.

`ERC20Airdrop2`'s missing `getBalance` also implemented (with tests) for unlocking.

### CODE REVIEW

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    /// @return balance The balance the user successfully claimed
+    /// @return withdrawableAmount The amount available to withdraw
+    function getBalance(address user)
+        public
+        view
+        returns (uint256 balance, uint256 withdrawableAmount)
+    {
+        balance = claimedAmount[user];
+        if (balance == 0) return (0, 0);
+        uint256 timeBasedAllowance = balance * uint64(block.timestamp - claimEnd) / withdrawalWindow;
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:84]: block.timestamp maybe super large if the user wait long...

Comment (dantaik) [reply packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:84]: we need a test for this contract

Comment (adaki2004) [reply packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:84]: > block.timestamp maybe super large if the user wait long...

Withdraw only allows to withdraw within the window. (Tho people could still query the readonly function of getBalance() so i’ll cap it.

Comment (adaki2004) [reply packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:84]: > we need a test for this contract

Added one testing the withdrawal part, but will add +1-2 extra more.

**Comment by dantaik** [comment]
Two questions:

1. If the USDC contract code must be deployed as-is on the byte-level, why don't we simply copy the byte-code into this project?
2. If the design is generic, we can/should test the design with a mock native token contract, not with a real one, which will avoid copying so many test contract files into this repo.

**Comment by adaki2004** [comment]
> Two questions:
> 
>     1. If the USDC contract code must be deployed as-is on the byte-level, why don't we simply copy the byte-code into this project?
> 
>     2. If the design is generic, we can/should test the design with a mock native token contract, not with a real one, which will avoid copying so many test contract files into this repo.

I think you meant to ask these questions in this PR https://github.com/taikoxyz/taiko-mono/pull/15250, no ?

But if we are here, my answers:
1. Yes, i agree this is they way, i just wanted a clean USDC contract which represents the actual state. We can copy the USDC bytecode and deploy it - on mainnet/testnet.
2. I would say yes and no at the same time.
Generic in terms of any kind of native assets could be plugged into the system.
Non-generic when it comes to each individual contracts - like USDC. It requires a 'translator' which maps the abbreviations (which in the USDC case is the burn function). So i wanted to E2E test a valid USDC contract, so that we are sure we know what we are doing, how we doing it, and it works on testnets and mainnets. I think it is definitely worth having +some extra files for safety- in the test folder, then figuring out live on testnet what is wrong.

**Comment by dantaik** [comment]
@adaki2004 May I ask you to rebased this PR with branch `proxy`?

File: packages/protocol/contracts/team/airdrop/MerkleClaimable.sol [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    "lib/openzeppelin-contracts-upgradeable/contracts/access/OwnableUpgradeable.sol";
+import { ReentrancyGuardUpgradeable } from
+    "lib/openzeppelin-contracts-upgradeable/contracts/security/ReentrancyGuardUpgradeable.sol";
+
+import { MerkleProofUpgradeable } from
+    "lib/openzeppelin-contracts-upgradeable/contracts/utils/cryptography/MerkleProofUpgradeable.sol";
+
+/// @title MerkleClaimable
+/// Contract for managing Taiko token airdrop for eligible users
+abstract contract MerkleClaimable is ReentrancyGuardUpgradeable, OwnableUpgradeable {
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/MerkleClaimable.sol:19]: Lets use EssentialContract as the base

File: packages/protocol/contracts/team/airdrop/MerkleClaimable.sol [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        uint64 _claimEnd,
+        bytes32 _merkleRoot
+    )
+        external
+        onlyOwner
+    {
+        _setConfig(_claimStart, _claimEnd, _merkleRoot);
+    }
+
+    function _init() internal {
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/MerkleClaimable.sol:77]: To follow open zepplin's naming conv, please rename it.

Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/MerkleClaimable.sol:77]:
```suggestion
    function _MerkleClaimable_init() internal {
```

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+        view
+        returns (uint256 balance, uint256 withdrawableAmount)
+    {
+        balance = claimedAmount[user];
+        // If balance is 0, or if the timestamp is above the (withdrawWindow + claimEnd) then there
+        // is no more balance and withdrawable amount
+        if (balance == 0 || uint64(block.timestamp) > (claimEnd + withdrawalWindow)) return (0, 0);
+        // Balance might be positive before end of claiming (claimEnd - if claimed already) but
+        // withdrawable is 0.
+        if (uint64(block.timestamp) < (claimEnd)) return (balance, 0);
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:89]:
```suggestion
        if (block.timestamp < claimEnd) return (balance, 0);
```

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    /// @return withdrawableAmount The amount available to withdraw
+    function getBalance(address user)
+        public
+        view
+        returns (uint256 balance, uint256 withdrawableAmount)
+    {
+        balance = claimedAmount[user];
+        // If balance is 0, or if the timestamp is above the (withdrawWindow + claimEnd) then there
+        // is no more balance and withdrawable amount
+        if (balance == 0 || uint64(block.timestamp) > (claimEnd + withdrawalWindow)) return (0, 0);
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:86]: If the time is way beyond the withdrawal window, then the user can withdraw all its unwithdrawn amount, not 0.

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        // Balance might be positive before end of claiming (claimEnd - if claimed already) but
+        // withdrawable is 0.
+        if (uint64(block.timestamp) < (claimEnd)) return (balance, 0);
+
+        uint256 timeBasedAllowance =
+            balance * (uint64(block.timestamp) - claimEnd) / withdrawalWindow;
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:92]: still buggy here.

**Review by dantaik — CHANGES_REQUESTED** [review]

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol (hunk 6 above)
Comment (adaki2004) [reply packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol]: Yep, thanks, corrected.

File: packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        returns (uint256 balance, uint256 withdrawableAmount)
+    {
+        balance = claimedAmount[user];
+        // If balance is 0 then there is no balance and withdrawable amount
+        if (balance == 0) return (0, 0);
+        // Balance might be positive before end of claiming (claimEnd - if claimed already) but
+        // withdrawable is 0.
+        if (block.timestamp < claimEnd) return (balance, 0);
+
+        // Hard cap timestamp - so range cannot go over - to get more allocation over time.
```
Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:91]: block.timestamp.min(claimEnd + withdrawalWindow)

Comment (dantaik) [inline packages/protocol/contracts/team/airdrop/ERC20Airdrop2.sol:91]: block.timestamp.min(claimEnd + withdrawalWindow)

**Review by dantaik — APPROVED** [review]

**Pull request merged**